adapters/huaweiads/huaweiads.go

```diff
@@ -0,0 +1,1320 @@
+				event = "complete"
+			case "playResume":
+				event = "resume"
+			case "playPause":
+				event = "pause"
+			case "soundClickOff":
+				event = "mute"
+			case "soundClickOn":
+				event = "unmute"
+			default:
+			}
+			if event != "" {
+				if event != "skip&closeLinear" {
+					trackingEvents.WriteString(getVastEventTrackingUrls(monitor.Url, event))
+				} else {
+					trackingEvents.WriteString(getVastEventTrackingUrls(monitor.Url, "skip&closeLinear"))
+				}
+			}
+		}
+	}
+
+	adm = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>" +
+		"<VAST version=\"3.0\">" +
+		"<Ad id=\"" + adId + "\"><InLine>" +
+		"<AdSystem>HuaweiAds</AdSystem>" +
+		"<AdTitle>" + adTitle + "</AdTitle>" +
+		errorTracking2Str + dspImpTracking2Str +
+		"<Creatives>" +
+		"<Creative adId=\"" + adId + "\" id=\"" + creativeId + "\">" +
+		"<Linear>" +
+		"<Duration>" + duration + "</Duration>" +
+		"<TrackingEvents>" + trackingEvents.String() + "</TrackingEvents>" +
+		"<VideoClicks>" +
+		"<ClickThrough><![CDATA[" + clickUrl + "]]></ClickThrough>" +
+		dspClickTracking2Str +
+		"</VideoClicks>" +
+		"<MediaFiles>" +
+		"<MediaFile delivery=\"progressive\" type=\"" + mime + "\" width=\"" + strconv.Itoa(int(adWidth)) + "\" " +
+		"height=\"" + strconv.Itoa(int(adHeight)) + "\" scalable=\"true\" maintainAspectRatio=\"true\"> " +
+		"<![CDATA[" + resourceUrl + "]]>" +
+		"</MediaFile>" +
+		"</MediaFiles>" +
+		"</Linear>" +
+		"</Creative>" +
+		"</Creatives>" +
+		"</InLine>" +
+		"</Ad>" +
+		"</VAST>"
+	return adm, adWidth, adHeight, nil
+}
+
+func getVastImpClickErrorTrackingUrls(urls []string, eventType string) (result string) {
+	var trackingUrls strings.Builder
+	for _, url := range urls {
+		if eventType == "click" {
+			trackingUrls.WriteString("<ClickTracking><![CDATA[" + url + "]]></ClickTracking>")
+		} else if eventType == "imp" {
+			trackingUrls.WriteString("<Impression><![CDATA[" + url + "]]></Impression>")
+		} else if eventType == "vastError" {
+			trackingUrls.WriteString("<Error><![CDATA[" + url + "&et=[ERRORCODE]]]></Error>")
+		}
+	}
+	return trackingUrls.String()
+}
+
+func getVastEventTrackingUrls(urls []string, eventType string) (result string) {
+	var trackingUrls strings.Builder
+	for _, eventUrl := range urls {
+		if eventType == "skip&closeLinear" {
+			trackingUrls.WriteString("<Tracking event=\"skip\"><![CDATA[" + eventUrl + "]]>" +
+				"</Tracking><Tracking event=\"closeLinear\"><![CDATA[" + eventUrl + "]]></Tracking>")
+		} else {
+			trackingUrls.WriteString("<Tracking event=\"" + eventType + "\"><![CDATA[" + eventUrl + "]]></Tracking>")
+		}
+	}
+	return trackingUrls.String()
+}
+
+// compute HmacSha256
+func computeHmacSha256(message string, signKey string) string {
+	h := hmac.New(sha256.New, []byte(signKey))
+	h.Write([]byte(message))
+	return hex.EncodeToString(h.Sum(nil))
+}
+
+// get digest authorization for request header, huawei ads ppsadx API
+func getDigestAuthorization(huaweiAdsImpExt *openrtb_ext.ExtImpHuaweiAds) string {
```

Comment:
Please add test coverage for computeHmacSha256 and getDigestAuthorization to ensure proper operation.

Reply:
This tester has already tested it, I don’t know how to add it, because the logic is based on the current time to obtain the results, it is dynamically generated

Reply:
What you can do is add  a field to your adapter that takes an interface with one method. The method would normally just call `time.Now()`, and return that, or your entire string value for `nonce` below. The `builder` function would populate the adapter with an empty struct with the intended method. In testing, you would either replace the interface with one that returns a constant value, or just build the adapter by hand with a constant value providing interface. That way in tests you will always be dealing with a known value for the time, rather than the `time.Now()` that you will be using in production.

This does slightly complicate your adapter, but makes it much more testable.

Reply:
Done, thank you @hhhjort